File: SCALA/OCP_OTHER_CODING_PROBS/DupStrings/DupStrings.scala
// Find duplicated strings (words) in an array of words. Print the words that recur, as well as the total
// number of times a word recurs at least once.

object DupStrings {

    def dupInfo(wl: Array[String]) {

        var wordOccurs:Map[String, Int] = Map()
        
        var index = 0

        for (index <- 0 to wl.length-1) {

            var thisWord = wl(index)

            var count = 0

            if ( wordOccurs.contains( thisWord )) {

                count = (wordOccurs.get(thisWord)).getOrElse(0)
            }

            count += 1

            wordOccurs += (thisWord -> count)
        }


        var repeats = 0

        for ( (word, count) <- wordOccurs) {

            if (count > 1) {

                println("%s : %d\n".format(word, count))

                repeats += 1;
            }
        }

        println( "Total words with 2+ occurences: %d\n".format(repeats ))
        println(wordOccurs)

         
    }

    //////////////////////////////////////////////////////////////////////////

    def main(args: Array[String]) {

        val testWords1 = Array("delhi", "bombay", "delhi", "dodballapur", "bombay", "calcutta", "bombay")

        dupInfo(testWords1)

    }
}

File: JS/OCP_OTHER_CODING_PROBS/DupStrings.js
// Find duplicated strings (words) in an array of words. Print the words that recur, as well as the total
// number of times a word recurs at least once.

class DupStrings {

    static dupInfo( wl ) {

        const wordOccurs = new Map();
        const wordSet = new Set();

        for (let index=0; index < wl.length; index++ ) {

            var thisWord = wl[index];

            if (wordOccurs[thisWord] != undefined) {

                var wcount = wordOccurs[thisWord];

                wordOccurs[thisWord] = (wcount + 1);
            }
            else {
                wordOccurs[thisWord] = 1;
            }

            wordSet.add(thisWord);
        }
        const wordAry = Array.from(wordSet);

        var repeats = 0;
        for (let index=0; index < wordAry.length; index++ ) {

            var word = wordAry[index];
            var occurs = wordOccurs[word];
            if (occurs > 1) {
                console.log(word);
                repeats += 1;
            }
        }
        console.log("Total words with 2+ occurences:" + repeats);
        console.log(wordOccurs);
         
    }
}

//////////////////////////////////////////////////////////////////////////

testWords1 = [ "delhi", "bombay", "delhi", "dodballapur", "bombay", "calcutta", "bombay"];

DupStrings.dupInfo(testWords1);
